### PR TITLE
Formatted image file data structure that breaks down by subdir

### PR DESCRIPTION
**Pull request by rdmark**
- New utility method for the web app, which sorts image files into dicts where the subdir is the key
- In the web ui, display each subdir in a table nested in a details tag.
- Allow for picking destination subdir when uploading files
- Style the expandable details blocks in the images table
- Add a check for ~ paths to the is_safe_path() utility method

### CODE REVIEW

File: python/web/src/web.py [hunk 1]
```diff
@@ -990,9 +990,10 @@ def upload_file():
         return make_response(auth["msg"], 403)
 
     destination = request.form.get("destination")
+    subdir = request.form.get("subdir")
     if destination == "disk_images":
         server_info = piscsi_cmd.get_server_info()
-        destination_dir = server_info["image_dir"]
+        destination_dir = server_info["image_dir"] + "/" + subdir
```
Comment (nucleogenic) [inline python/web/src/web.py:996]: Could this path be manipulated by a malicious actor to write to unauthorised locations?

Comment (rdmark) [reply python/web/src/web.py:996]: Right, security. That's a thing. :)

Comment (rdmark) [reply python/web/src/web.py]: We already have the handy `is_safe_path()` utility method. Easy fix!

File: python/web/src/templates/index.html [hunk 2]
```diff
@@ -209,26 +209,33 @@
...
+
+<div>
+{% for subdir, group in formatted_image_files.items() %}
+{% if subdir != "/" %}
+<details id="subdir">
+    <summary class="dirname">
+        images/{{ subdir }}
+    </summary>
+{% endif %}
 <table id="images" border="black" cellpadding="3" summary="List of files in the image directory">
```
Comment (nucleogenic) [inline python/web/src/templates/index.html:226]: I would also recommend giving tables a small amount of vertical margin to separate them from the directory name above/below (e.g. `margin: 0.5rem 0;`)

image

**Review by nucleogenic — APPROVED** [review]
Approving as functionally complete, comments optional.

File: python/web/src/templates/index.html [hunk 3]
```diff
@@ -209,26 +209,33 @@
...
+{% else %}
+
+<div>
+{% for subdir, group in formatted_image_files.items() %}
+{% if subdir != "/" %}
+<details id="subdir">
+    <summary class="dirname">
+        images/{{ subdir }}
+    </summary>
+{% endif %}
```
[the remark this answers is not in the document]
Comment (rdmark) [reply python/web/src/templates/index.html:225]: That's a pretty neat idea for consistency. Didn't know about the `open` property for details.

File: python/web/src/templates/index.html (hunk 2 above)
Comment (rdmark) [reply python/web/src/templates/index.html:226]: The master of UX!

[the remark this answers is not in the document]
Comment (rdmark) [reply python/web/src/templates/index.html:226]: Good call.

File: python/web/src/web_utils.py [hunk 4]
```diff
@@ -146,6 +147,32 @@ def get_image_description(file_suffix):
...
+            subdir = subdir_path[0]
+            if subdir in subdir_image_files.keys():
+                subdir_image_files[subdir].append(image)
+            else:
+                subdir_image_files[subdir] = [image]
+        else:
+            root_image_files.append(image)
+
+    formatted_image_files = dict(sorted(subdir_image_files.items()))
+    formatted_image_files["/"] = root_image_files
```
[the remark this answers is not in the document]
Comment (rdmark) [reply python/web/src/web_utils.py:172]: Are you sure this is really the case? This dict gets its data from FileCmds.list_images() which in turn gets it from the DEFAULT_IMAGE_FILES_INFO protobuf request, which promptly ignores empty dirs (and files <512 bytes) as per piscsi backend image file validation logic. When I inspected the data that comes through only populated dirs are represented. 

I could verify this with dynamic testing as well. Empty dirs are not represented in the UI.

There might be a corner case I'm overlooking though so please guide me how to reproduce. :)

[the remark this answers is not in the document]
Comment (nucleogenic) [reply python/web/src/web_utils.py:172]: I believe this was caused by `root_image_files` first being set to an empty list, then `formatted_image_files["/"] = root_image_files` adds the empty entry to the result. If that doesn't instantly make sense, I'll jump on my workstation and do a proper retest/reproduction instructions.

**Comment by nucleogenic** [comment]
All looking good!

**Pull request merged**